pysindy/optimizers/constrained_stlsq.py

```diff
@@ -0,0 +1,169 @@
+import warnings
+
+import numpy as np
+from sklearn.exceptions import ConvergenceWarning
+from sklearn.linear_model import ridge_regression
+
+from pysindy.optimizers import BaseOptimizer
+
+
+class ConstrainedSTLSQ(BaseOptimizer):
+    """
+    Sequentially thresholded least squares algorithm.
+
+    Attempts to minimize the objective function
+    ||y - Xw||^2_2 + alpha * ||w||^2_2 by iteratively performing
+    least squares and masking out elements of the weight that are
+    below a given threshold.
+
+    subject to  Cw = d (equality constraints)
+
+    Parameters
+    ----------
+    threshold : float, optional (default 0.1)
+        Minimum magnitude for a coefficient in the weight vector.
+        Coefficients with magnitude below the threshold are set
+        to zero.
+
+    alpha : float, optional (default 0)
+        Optional L2 (ridge) regularization on the weight vector.
+
+    max_iter : int, optional (default 20)
+        Maximum iterations of the optimization algorithm.
+
+    ridge_kw : dict, optional
+        Optional keyword arguments to pass to the ridge regression.
+
+    initial_guess : 2D numpy array of floats (default None)
+        If user does not pass this, the initial guess for the
+        optimization is a naive lstsq (see below). If passes,
+        the optimization starts with this matrix as the initial
+        starting point.
+
+    Attributes
+    ----------
+    coef_ : array, shape (n_features,) or (n_targets, n_features)
+        Weight vector(s)
+
+    ind_ : array, shape (n_features,) or (n_targets, n_features)
+        Array of 0s and 1s indicating which coefficients of the
+        weight vector have not been masked out.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy.integrate import odeint
+    >>> from pysindy import SINDy
+    >>> from pysindy.optimizers import ConstrainedSTLSQ
+    >>> lorenz = lambda z,t : [10*(z[1] - z[0]),
+    >>>                        z[0]*(28 - z[2]) - z[1],
+    >>>                        z[0]*z[1] - 8/3*z[2]]
+    >>> t = np.arange(0,2,.002)
+    >>> x = odeint(lorenz, [-8,8,27], t)
+    >>> opt = ConstrainedSTLSQ(threshold=.1, alpha=.5)
+    >>> model = SINDy(optimizer=opt)
+    >>> model.fit(x, t=t[1]-t[0])
+    >>> model.print()
+    x0' = -9.999 1 + 9.999 x0
+    x1' = 27.984 1 + -0.996 x0 + -1.000 1 x1
+    x2' = -2.666 x1 + 1.000 1 x0
+    """
+
+    def __init__(
```

Comment:
Where do the constraints come in? I see that the `ConstrainedSR3` class has some additional parameters specifying the constraints, but I don't see them for this class.